src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs

```diff
@@ -0,0 +1,197 @@
+                    null,
+                    (state, ex) => state.ToString());
+            }
+        }
+
+        public static void LogRequestEnd(this ILogger logger, HttpResponseMessage response, TimeSpan duration, Func<string, bool> shouldRedactHeaderValue)
+        {
+            _requestEnd(logger, duration.TotalMilliseconds, (int)response.StatusCode, null);
+
+            if (logger.IsEnabled(LogLevel.Trace))
+            {
+                logger.Log(
+                    LogLevel.Trace,
+                    EventIds.ResponseHeader,
+                    new HttpHeadersLogValue(HttpHeadersLogValue.Kind.Response, response.Headers, response.Content?.Headers, shouldRedactHeaderValue),
+                    null,
+                    (state, ex) => state.ToString());
+            }
+        }
+
+        public static IDisposable? BeginRequestPipelineScope(this ILogger logger, HttpRequestMessage request)
+        {
+            return _beginRequestPipelineScope(logger, request.Method, GetRedactedUriString(request.RequestUri));
+        }
+
+        public static void LogRequestPipelineStart(this ILogger logger, HttpRequestMessage request, Func<string, bool> shouldRedactHeaderValue)
+        {
+            _requestPipelineStart(logger, request.Method, GetRedactedUriString(request.RequestUri), null);
+
+            if (logger.IsEnabled(LogLevel.Trace))
+            {
+                logger.Log(
+                    LogLevel.Trace,
+                    EventIds.RequestPipelineRequestHeader,
+                    new HttpHeadersLogValue(HttpHeadersLogValue.Kind.Request, request.Headers, request.Content?.Headers, shouldRedactHeaderValue),
+                    null,
+                    (state, ex) => state.ToString());
+            }
+        }
+
+        public static void LogRequestPipelineEnd(this ILogger logger, HttpResponseMessage response, TimeSpan duration, Func<string, bool> shouldRedactHeaderValue)
+        {
+            _requestPipelineEnd(logger, duration.TotalMilliseconds, (int)response.StatusCode, null);
+
+            if (logger.IsEnabled(LogLevel.Trace))
+            {
+                logger.Log(
+                    LogLevel.Trace,
+                    EventIds.RequestPipelineResponseHeader,
+                    new HttpHeadersLogValue(HttpHeadersLogValue.Kind.Response, response.Headers, response.Content?.Headers, shouldRedactHeaderValue),
+                    null,
+                    (state, ex) => state.ToString());
+            }
+        }
+
+        internal static string? GetRedactedUriString(Uri? uri)
+        {
+            if (uri is null)
+            {
+                return null;
+            }
+
+            string uriString = uri.IsAbsoluteUri
+                ? uri.GetComponents(UriComponents.AbsoluteUri & ~UriComponents.UserInfo, UriFormat.UriEscaped)
+                : uri.ToString();
```

Comment:
We also have to remove userinfo. In order to also remove it from relative Uri-s we would need to duplicate a non-trivial amount of Uri parsing code in the package. Given that the prence of relative Uri-s here is an edge-case  (one would need to use `LoggingHttpMessageHandler` manually with `HttpMessageInvoker`), I'm not sure if we should bother doing that.

@blowdart @MihaZupan thoughts?

Marking this as `NO-MERGE` until this is resolved.

Reply:
Alternatively, we can omit logging relative Uri-s, IMHO that would be the simplest thing to deal with this edge-case.

Reply:
I pushed a change that redacts relative URIs unconditionally, PTAL.

Reply:
People still use username/password in URIs? It's long deprecated now and most modern servers don't even do anything with it :)

Reply:
> It's long deprecated now

But isn't only `username:password` combination deprecated? 🤔🤔🤔 and `usename`-only still can be used and should be supported? 🤔 

(I judge only by what I've seen in the RFC though, so I might be wrong)

<details>

https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.1

I understand the following _only_ applies to the "user:password" format:

```
   Use of the format "user:password" in the userinfo field is
   deprecated.  Applications should not render as clear text any data
   after the first colon (":") character found within a userinfo
   subcomponent unless the data after the colon is the empty string
   (indicating no password).  Applications may choose to ignore or
   reject such data when it is received as part of a reference and
   should reject the storage of such data in unencrypted form.  The
   passing of authentication information in clear text has proven to be
   a security risk in almost every case where it has been used.
```

And the following implies that `usename`-only is ok if the user wants to see it:

```
   Applications that render a URI for the sake of user feedback, such as
   in graphical hypertext browsing, should render userinfo in a way that
   is distinguished from the rest of a URI, when feasible.
```

</details>

Reply:
So talking to Anton in teams I believe the sticking point here is for URIs with relative schemes, eg. "//user:pass@example/foo/bar".

But does the http client support those? They really don't make sense outside of a containing web page where the protocol is taken from the page that contains such links?

Reply:
I don't think that's the case. According to the [4.2 Relative Reference](https://datatracker.ietf.org/doc/html/rfc3986#section-4.2).

> A relative reference takes advantage of the hierarchical syntax (Section 1.2.3) to express a URI reference relative to the name space of another hierarchical URI.

Which seems to be the definition of what we call relative Uri. I'm still having trouble to imagine how can relative Uri-s contain UserInfo in practice. @MihaZupan do you have any insight?

Reply:
> do you have any insight?

Only that when you first asked me about it, I had to double-check that this is a thing, and that `System.Uri` supports it.
You can only supply the user info when also supplying the authority, and that is only possible with the `"//" authority path-abempty` syntax, which even the RFC calls out as rare
> A relative reference that begins with two slash characters is termed a network-path reference; such references are rarely used.

This is also a format where it may be ambiguous as to whether you mean a relative network-path reference vs. an absolute UNC path depending on whether you allow `Uri` to use `UriKind.Relative` or not.
```c#
Console.WriteLine(new Uri("//domain/path", UriKind.Absolute).IsAbsoluteUri);
Console.WriteLine(new Uri("//domain/path", UriKind.Relative).IsAbsoluteUri);
```

---

> But does the http client support those?

Our code that sends the request (SocketsHttpHandler) only cares about absolute Uris.

But technically a user can create a custom `HttpMessageHandler` chain that accepts a relative Uri and then replaces it with an absolute one sometime later before calling into `SocketsHttpHandler`.
And inside this chain they _could_ call into `LoggingHttpMessageHandler` before they've replaced the Uri with an absolute one.

---

TL;DR; IMO it seems like a pretty obscure thing for someone to do, in which case they should just... not do that :)

Reply:
I guess someone could be writing a .NET based browser and wants protocol relative, rather than just path relative. What a mess. URIs with UNC formats shouldn't accept username/passwords though, which presumably would help disambiguate.

So, given the format which requires user:pass@ in the first segment after an optional scheme, can you not just pattern match on that and strip the segment?

Reply:
I think the simplest thing is to stick with the strategy of redacting relative Uri-s entirely. If anyone will be bitten by this, e557330 now enables opting out from redaction entirely & they can open an issue for us to revisit this decision.

Scream if you disagree :)